EIPS/eip-7399.md

```diff
@@ -0,0 +1,217 @@
+ - Some lenders offer a single entry point into the protocol regardless of whether you're buying, selling, depositing or chaining them together as a flash loan, whereas other protocols offer discrete entry points.
+ - Some lenders allow to flash mint any amount of their native asset without charging a fee, effectively allowing flash loans bounded by computational constraints instead of asset ownership constraints.
+
+## Specification
+
+A flash loan integrates two smart contracts using a callback pattern. These are called the `lender` and the `callback receiver` in this EIP.
+
+In the same flash loan, assets are transferred to the `loan receiver` before the callback and expected in the `payment receiver` at the end of the callback. An additional amount of assets, called a `fee`, MIGHT be also required to the present in the `payment receiver` at the end of the callback.
+
+### Lender Specification
+
+A `lender` MUST implement the IERCXXXXFlashLender interface.
+
+```
+// SPDX-License-Identifier: MIT
+pragma solidity >=0.6.0 <0.9.0;
+import { IERC20 } from "./IERC20.sol";
+
+interface IERC3156PPFlashLender {
+
+    /// @dev The fee to be charged for a given loan.
+    /// @param asset The loan currency.
+    /// @param amount The amount of assets lent.
+    /// @return The amount of `asset` to be charged for the loan, on top of the returned principal. Returns type(uint256).max if the loan is not possible.
+    function flashFee(
+        IERC20 asset,
+        uint256 amount
+    ) external view returns (uint256);
+
+    /// @dev Use the aggregator to serve an ERC3156++ flash loan.
+    /// @param loanReceiver The address receiving the assets in the flash loan
+    /// @param asset The asset to be loaned
+    /// @param amount The amount to loaned
+    /// @param data The ABI encoded user data
+    /// @param callback The address and signature of the callback function
+    /// @return result ABI encoded result of the callback
+    function flash(
+        address loanReceiver,
+        IERC20 asset,
+        uint256 amount,
+        bytes calldata data,
+        /// @dev callback. This is a concatenation of (address, bytes4), where the address is the callback receiver, and the bytes4 is the signature of callback function.
+        /// @param initiator The address that called this function
+        /// @param paymentReceiver The address that needs to receive the amount plus fee at the end of the callback
+        /// @param asset The asset to be loaned
+        /// @param amount The amount to loaned
+        /// @param fee The fee to be paid
+        /// @param data The ABI encoded data to be passed to the callback
+        /// @return result ABI encoded result of the callback
+        function(address, address, IERC20, uint256, uint256, bytes memory) external returns (bytes memory) callback
```

Comment:
Is this valid syntax? I've never seen it before.

Reply:
yup
https://docs.soliditylang.org/en/v0.8.20/types.html#function-types
https://twitter.com/PaulRBerg/status/1609917508223475712?s=20

Reply:
I've added some extra comments in the EIP explaining how it works.